File: Decorator_Pattern/dp.cpp
#include <algorithm>
#include <iostream>
#include <memory>
#include <vector>

class Bverage {
   protected:
    std::string desc = "Unknown Beverage";

   public:
    std::string getDesc() {
        return this->desc;
    }
    virtual double cost() const = 0;
};

class CondimentDecorator : public Bverage {
   public:
    virtual std::string getDesc() = 0;
};

class Expresso : public Bverage {
   public:
    explicit Expresso() {
        this->desc = "Expresso Coffee";
    }

    double cost() const override {
        return 1.99;
    }
};

class HouseBlend : public Bverage {
   public:
    explicit HouseBlend() {
        this->desc = "House Blend Coffee";
    }

    double cost() const override {
        return 0.89;
    }
};

class Mocha : public CondimentDecorator {
    std::unique_ptr<Bverage> beverage;

   public:
    explicit Mocha(std::unique_ptr<Bverage> _beverage)
        : beverage(std::move(_beverage)) {}

    std::string getDesc() {
        return beverage->getDesc() + "Mocha";
    }

    double cost() const override {
        return 0.20 + beverage->cost();
    }
};

class Soy : public CondimentDecorator {
    std::unique_ptr<Bverage> beverage;

   public:
    explicit Soy(std::unique_ptr<Bverage> _beverage)
        : beverage(std::move(_beverage)) {}

    std::string getDesc() {
        return beverage->getDesc() + "Soy";
    }

    double cost() const override {
        return 0.15 + beverage->cost();
    }
};

class Whip : public CondimentDecorator {
    std::unique_ptr<Bverage> beverage;

   public:
    explicit Whip(std::unique_ptr<Bverage> _beverage)
        : beverage(std::move(_beverage)) {}

    std::string getDesc() {
        return beverage->getDesc() + "Whip";
    }

    double cost() const override {
        return 0.10 + beverage->cost();
    }
};

int main() {
	std::unique_ptr<Bverage> beverage = std::make_unique<Expresso>();
	std::cout<<"Beverage: "<<beverage->getDesc()<<" $"<<beverage->cost()<<"\n";

	std::unique_ptr<Bverage> beverage2 = std::make_unique<HouseBlend>();
	beverage2 = std::make_unique<Mocha>(std::move(beverage2));
	beverage2 = std::make_unique<Mocha>(std::move(beverage2));
	beverage2 = std::make_unique<Whip>(std::move(beverage2));
	std::cout<<"Beverage2: "<<beverage2->getDesc()<<" $"<<beverage2->cost()<<"\n";


}

File: Decorator_Pattern/dp.py
from abc import ABC, abstractmethod

class Beverage(ABC):
    def __init__(self) -> None:
        super().__init__()
        self.desc = "Unknown beverage"

    def get_desc(self) -> str: return self.desc

    @abstractmethod
    def cost(self) -> float: pass

class CondimentDecorator(Beverage, ABC):
    def __call__(self) -> None:
        self.beverage
    
    @abstractmethod
    def get_desc(self) -> str: pass

class Expresso(Beverage):
    def __init__(self) -> None:
        super().__init__()
        self.desc = 'Expresso'

    def cost(self) -> float:
        return 1.99

class HouseBlend(Beverage):
    def __init__(self) -> None:
        super().__init__()
        self.desc = 'HouseBlend'

    def cost(self) -> float:
        return 1.99

class Mocha(CondimentDecorator):
    def __init__(self, beverage: Beverage) -> None:
        super().__init__()
        self.beverage = beverage

    def get_desc(self) -> str:
        return self.beverage.get_desc() + 'Mocha'

    def cost(self) -> float:
        return 0.20 + self.beverage.cost()

class Soy(CondimentDecorator):
    def __init__(self, beverage: Beverage) -> None:
        super().__init__()
        self.beverage = beverage

    def get_desc(self) -> str:
        return self.beverage.get_desc() + 'Soy'

    def cost(self) -> float:
        return 0.15 + self.beverage.cost()

def main():
    beverage = Expresso()
    print(f'{beverage.get_desc()} ${beverage.cost()}')

    b2 = HouseBlend()
    b2 = Mocha(b2)
    b2 = Soy(b2)
    print(f'{b2.get_desc()} ${b2.cost()}')


if __name__ == '__main__':
    main()
    